viskores/cont/CubicHermiteSpline.cxx

```diff
@@ -0,0 +1,145 @@
+
+#include <viskores/BinaryOperators.h>
+#include <viskores/VectorAnalysis.h>
+#include <viskores/cont/Algorithm.h>
+#include <viskores/cont/ArrayHandleConstant.h>
+#include <viskores/cont/CubicHermiteSpline.h>
+#include <viskores/exec/CubicHermiteSpline.h>
+#include <viskores/worklet/WorkletMapField.h>
+
+namespace viskores
+{
+namespace cont
+{
+namespace
+{
+struct CalcNeighborDistanceWorklet : public viskores::worklet::WorkletMapField
+{
+  using ControlSignature = void(FieldOut, WholeArrayIn);
+  using ExecutionSignature = void(InputIndex, _1, _2);
+
+  template <typename ArrayType>
+  VISKORES_EXEC void operator()(const viskores::Id& idx,
+                                viskores::FloatDefault& val,
+                                const ArrayType& data) const
+  {
+    if (idx == 0)
+      val = 0.0;
+    else
+      val = viskores::Magnitude(data.Get(idx) - data.Get(idx - 1));
+  }
+};
+
+struct CalcTangentsWorklet : public viskores::worklet::WorkletMapField
+{
+  CalcTangentsWorklet(const viskores::Id& numPoints)
+    : NumPoints(numPoints)
+  {
+  }
+
+  using ControlSignature = void(FieldOut, WholeArrayIn, WholeArrayIn);
+  using ExecutionSignature = void(InputIndex, _1, _2, _3);
+
+  template <typename TangentArrayType, typename PointArrayType, typename KnotArrayType>
+  VISKORES_EXEC void operator()(const viskores::Id& idx,
+                                TangentArrayType& tangent,
+                                const PointArrayType& points,
+                                const KnotArrayType& knots) const
+  {
+    viskores::Id idx0 = viskores::Max(viskores::Id(0), idx - 1);
+    viskores::Id idx1 = viskores::Min(NumPoints - 1, idx + 1);
+
+    auto dX = points.Get(idx1) - points.Get(idx0);
+    auto dT = knots.Get(idx1) - knots.Get(idx0);
+
+    tangent = dX / dT;
+  }
+
+  viskores::Id NumPoints;
+};
+} //anonymous namespace
+
+VISKORES_CONT viskores::Range CubicHermiteSpline::GetParametricRange()
+{
+  auto n = this->Knots.GetNumberOfValues();
+  if (n == 0)
+  {
+    this->ComputeKnots();
+    n = this->Knots.GetNumberOfValues();
+  }
+  const auto ids = viskores::cont::make_ArrayHandle<viskores::Id>({ 0, n - 1 });
+  const std::vector<viskores::FloatDefault> output =
+    viskores::cont::ArrayGetValues(ids, this->Knots);
+  return { output[0], output[1] };
+}
+
+VISKORES_CONT void CubicHermiteSpline::ComputeKnots()
+{
+  viskores::Id n = this->Data.GetNumberOfValues();
+  this->Knots.Allocate(n);
+
+  viskores::cont::Invoker invoker;
+
+  //uses chord length parameterization.
+  invoker(CalcNeighborDistanceWorklet{}, this->Knots, this->Data);
+  viskores::FloatDefault sum = viskores::cont::Algorithm::ScanInclusive(this->Knots, this->Knots);
+
+  if (sum == 0.0)
+    throw std::invalid_argument("Error: accumulated distance between data is zero.");
+
+  auto divideBy =
+    viskores::cont::make_ArrayHandleConstant(1.0 / sum, this->Knots.GetNumberOfValues());
```

Comment:
The following line is leading to a warning deep in the call to `Transform`. I'm pretty sure it is because this constant array is set to contain `viskores::Float64` and that is being computed with arrays of type `viskores::DefaultFloat` (which for this build is 32-bit float).

```suggestion
  auto divideBy =
    viskores::cont::make_ArrayHandleConstant(static_cast<viskores::FloatDefault>(1.0 / sum), this->Knots.GetNumberOfValues());
```